### PR TITLE
sdr-dsp: FIR tap generation

### LINKED ISSUE #4: sdr-dsp: FIR tap generation
## Description

Implement FIR filter tap/coefficient generation.

## C++ Reference

- `core/src/dsp/taps/` -- low_pass, band_pass, high_pass, root_raised_cosine
- `core/src/dsp/taps/estimate_tap_count.h` -- tap count estimation from transition bandwidth

## Deliverables

- `sdr-dsp/src/taps.rs`: low_pass, band_pass, high_pass tap generators using windowed sinc method
- Tap count estimation from transition bandwidth and attenuation
- Root raised cosine tap generation
- Unit tests verifying filter responses at key frequencies

### PR DESCRIPTION
**Pull request by jasonherald**
## Summary

- Implement `estimate_tap_count` for automatic tap sizing from transition width
- Implement `low_pass`, `high_pass`, `band_pass` using windowed sinc with Nuttall window
- Implement `root_raised_cosine` with standard RRC formula (special cases at t=0 and limits)
- Core `windowed_sinc` generator with pluggable window functions
- All functions return `Vec<f32>` tap vectors

## Test plan

- [x] 30 total tests passing (`cargo test -p sdr-dsp`) — 11 new tap tests
- [x] `cargo clippy --all-targets --workspace -- -D warnings` clean
- [x] `cargo fmt --all -- --check` clean
- [x] Lowpass: symmetric taps, odd count option, DC gain ~1.0
- [x] Highpass: DC gain ~0 (rejects DC)
- [x] Bandpass: DC gain ~0, invalid range panics
- [x] RRC: symmetric taps, peak at center

Closes #4

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a DSP taps module exposing FIR tap generation for low-pass, high-pass, band-pass, and root-raised-cosine filters.
  * Added tap-count estimation from transition width and sample rate with optional odd-length enforcement, input validation, and allocation limits.

* **Tests**
  * Added unit tests covering tap-count estimation, parameter validation/error handling, symmetry and gain properties, RRC limit-case handling, and non-degenerate outputs.

### CODE REVIEW

File: crates/sdr-dsp/src/taps.rs [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+/// Tap count estimation scaling factor (from SDR++ `estimateTapCount`).
+const TAP_COUNT_FACTOR: f64 = 3.8;
+
+/// Estimate the number of FIR filter taps needed for a given transition width.
+///
+/// Ports SDR++ `dsp::taps::estimateTapCount`.
+/// Formula: `count = 3.8 * sample_rate / transition_width`
+#[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]
+pub fn estimate_tap_count(transition_width: f64, sample_rate: f64) -> usize {
+    (TAP_COUNT_FACTOR * sample_rate / transition_width) as usize
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs]: Fixed — added validate_positive_finite guard. All public functions now return Result<Vec<f32>, DspError>.

File: crates/sdr-dsp/src/taps.rs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+/// - `odd_tap_count`: if true, ensures an odd number of taps
+pub fn high_pass(
+    cutoff: f64,
+    transition_width: f64,
+    sample_rate: f64,
+    odd_tap_count: bool,
+) -> Vec<f32> {
+    let mut count = estimate_tap_count(transition_width, sample_rate);
+    if odd_tap_count && count.is_multiple_of(2) {
+        count += 1;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs]: Fixed — high_pass always forces odd tap count now (Type I FIR). Removed the odd_tap_count parameter.

File: crates/sdr-dsp/src/taps.rs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    band_start: f64,
+    band_stop: f64,
+    transition_width: f64,
+    sample_rate: f64,
+    odd_tap_count: bool,
+) -> Vec<f32> {
+    assert!(
+        band_stop > band_start,
+        "band_stop must be greater than band_start"
+    );
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs]: Fixed — returns DspError::InvalidParameter instead of panicking.

File: crates/sdr-dsp/src/taps.rs [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    let offset_omega = math::hz_to_rads(f64::midpoint(band_start, band_stop), sample_rate);
+    let mut count = estimate_tap_count(transition_width, sample_rate);
+    if odd_tap_count && count.is_multiple_of(2) {
+        count += 1;
+    }
+    let bandwidth = (band_stop - band_start) / 2.0;
+    windowed_sinc(count, bandwidth, sample_rate, |n, big_n| {
+        #[allow(clippy::cast_possible_truncation)]
+        let modulation = 2.0 * (offset_omega * n).cos();
+        modulation * window::nuttall(n, big_n)
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs]: Fixed — band_pass now computes modulation using the centered tap coordinate (t = i - half + 0.5) directly instead of relying on the window closure parameter. Symmetry test added and passing.

File: crates/sdr-dsp/src/taps.rs [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+///
+/// - `count`: number of taps
+/// - `beta`: roll-off factor (0.0 to 1.0)
+/// - `symbol_rate`: symbol rate in symbols/second
+/// - `sample_rate`: sample rate in Hz
+#[allow(clippy::cast_precision_loss, clippy::cast_possible_truncation)]
+pub fn root_raised_cosine(count: usize, beta: f64, symbol_rate: f64, sample_rate: f64) -> Vec<f32> {
+    let ts = sample_rate / symbol_rate;
+    let half = count as f64 / 2.0;
+    let limit = ts / (4.0 * beta);
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs]: Fixed — added validation for count > 0, beta in 0..=1, and positive/finite rates. Returns DspError on invalid input.

File: crates/sdr-dsp/src/taps.rs [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    }
+
+    #[test]
+    fn test_band_pass_basic() {
+        let taps = band_pass(5_000.0, 10_000.0, 1_000.0, TEST_SAMPLE_RATE, false);
+        assert!(!taps.is_empty());
+        // Bandpass taps sum should be near zero (rejects DC)
+        let sum: f32 = taps.iter().sum();
+        assert!(sum.abs() < 0.2, "BP DC gain should be ~0, got {sum}");
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs:333]: Fixed — added assert_symmetric helper and symmetry tests for high_pass and band_pass.

File: crates/sdr-dsp/src/taps.rs [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+    transition_width: f64,
+    sample_rate: f64,
+    odd_tap_count: bool,
+) -> Result<Vec<f32>, DspError> {
+    validate_positive_finite(cutoff, "cutoff")?;
+    let mut count = estimate_tap_count(transition_width, sample_rate)?;
+    if odd_tap_count && count.is_multiple_of(2) {
+        count += 1;
+    }
+    Ok(windowed_sinc(count, cutoff, sample_rate, window::nuttall))
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs:73]: Fixed — low_pass now rejects cutoff >= Nyquist with a descriptive error.

File: crates/sdr-dsp/src/taps.rs [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+) -> Result<Vec<f32>, DspError> {
+    validate_positive_finite(cutoff, "cutoff")?;
+    let mut count = estimate_tap_count(transition_width, sample_rate)?;
+    // Always force odd tap count for highpass (Type I FIR avoids Nyquist zero)
+    if count.is_multiple_of(2) {
+        count += 1;
+    }
+    Ok(windowed_sinc(
+        count,
+        (sample_rate / 2.0) - cutoff,
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs:104]: Fixed — high_pass now validates cutoff < Nyquist before computing the effective lowpass cutoff.

File: crates/sdr-dsp/src/taps.rs [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+    sample_rate: f64,
+    odd_tap_count: bool,
+) -> Result<Vec<f32>, DspError> {
+    if band_stop <= band_start {
+        return Err(DspError::InvalidParameter(
+            "band_stop must be greater than band_start".to_string(),
+        ));
+    }
+    validate_positive_finite(band_start, "band_start")?;
+    validate_positive_finite(sample_rate, "sample_rate")?;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs:142]: Fixed — added validate_positive_finite for band_stop alongside band_start.

File: crates/sdr-dsp/src/taps.rs [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+    let taps = (0..count)
+        .map(|i| {
+            let t = i as f64 - half + 0.5;
+            let val = if t == 0.0 {
+                (1.0 + beta * (4.0 / PI - 1.0)) / ts
+            } else if (t - limit).abs() < 1e-12 || (t + limit).abs() < 1e-12 {
+                let sin_term = (PI / (4.0 * beta)).sin();
+                let cos_term = (PI / (4.0 * beta)).cos();
+                ((1.0 + 2.0 / PI) * sin_term + (1.0 - 2.0 / PI) * cos_term) * beta
+                    / (ts * core::f64::consts::SQRT_2)
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs:218]: Fixed — extracted to RRC_SINGULARITY_EPS named constant.

File: crates/sdr-dsp/src/taps.rs [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+    }
+
+    #[test]
+    fn test_estimate_tap_count_invalid() {
+        assert!(estimate_tap_count(0.0, TEST_SAMPLE_RATE).is_err());
+        assert!(estimate_tap_count(-1.0, TEST_SAMPLE_RATE).is_err());
+        assert!(estimate_tap_count(1_000.0, 0.0).is_err());
+        assert!(estimate_tap_count(f64::NAN, TEST_SAMPLE_RATE).is_err());
+        assert!(estimate_tap_count(f64::INFINITY, TEST_SAMPLE_RATE).is_err());
+    }
```
Comment (jasonherald) [inline crates/sdr-dsp/src/taps.rs:288]: Fixed — added f64::NEG_INFINITY to invalid parameter test coverage.

File: crates/sdr-dsp/src/taps.rs [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+///
+/// # Errors
+///
+/// Returns `DspError::InvalidParameter` if `transition_width` or `sample_rate`
+/// are non-positive or non-finite.
+#[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]
+pub fn estimate_tap_count(transition_width: f64, sample_rate: f64) -> Result<usize, DspError> {
+    validate_positive_finite(transition_width, "transition_width")?;
+    validate_positive_finite(sample_rate, "sample_rate")?;
+    Ok((TAP_COUNT_FACTOR * sample_rate / transition_width) as usize)
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs]: Fixed — added MAX_TAP_COUNT (1M) upper bound guard on estimate_tap_count to prevent unreasonable allocations from edge-case inputs.

File: crates/sdr-dsp/src/taps.rs [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+        assert_eq!(peak, center, "peak should be at center tap");
+    }
+
+    #[test]
+    fn test_root_raised_cosine_invalid() {
+        assert!(root_raised_cosine(0, 0.35, 9600.0, TEST_SAMPLE_RATE).is_err());
+        assert!(root_raised_cosine(65, 1.5, 9600.0, TEST_SAMPLE_RATE).is_err());
+        assert!(root_raised_cosine(65, -0.1, 9600.0, TEST_SAMPLE_RATE).is_err());
+        assert!(root_raised_cosine(65, 0.35, 0.0, TEST_SAMPLE_RATE).is_err());
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs:394]: Fixed — added test_root_raised_cosine_limit_branch with beta=0.625 that lands t=±2.0 exactly on the singularity. Verifies taps are finite, non-zero, and symmetric.

File: crates/sdr-dsp/src/taps.rs [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+    validate_positive_finite(band_start, "band_start")?;
+    validate_positive_finite(band_stop, "band_stop")?;
+    validate_positive_finite(sample_rate, "sample_rate")?;
+
+    let offset_omega = math::hz_to_rads(f64::midpoint(band_start, band_stop), sample_rate);
+    let mut count = estimate_tap_count(transition_width, sample_rate)?;
+    if odd_tap_count && count.is_multiple_of(2) {
+        count += 1;
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs:154]: Fixed — added Nyquist boundary check for band_stop. Rejects band_stop >= sample_rate/2.

File: crates/sdr-dsp/src/taps.rs [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+    let nyquist = sample_rate / 2.0;
+    if cutoff >= nyquist {
+        return Err(DspError::InvalidParameter(format!(
+            "cutoff ({cutoff}) must be less than Nyquist ({nyquist})"
+        )));
+    }
+    let mut count = estimate_tap_count(transition_width, sample_rate)?;
+    if odd_tap_count && count.is_multiple_of(2) {
+        count += 1;
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs:72]: Acknowledged — we're building on nightly and edition 2024 (no explicit rust-version set in workspace). is_multiple_of is available. If we need to pin an MSRV in the future we'll address this then.

File: crates/sdr-dsp/src/taps.rs [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+pub fn estimate_tap_count(transition_width: f64, sample_rate: f64) -> Result<usize, DspError> {
+    validate_positive_finite(transition_width, "transition_width")?;
+    validate_positive_finite(sample_rate, "sample_rate")?;
+    let count = (TAP_COUNT_FACTOR * sample_rate / transition_width) as usize;
+    if count > MAX_TAP_COUNT {
+        return Err(DspError::InvalidParameter(format!(
+            "estimated tap count ({count}) exceeds maximum ({MAX_TAP_COUNT})"
+        )));
+    }
+    Ok(count)
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs:46]: Fixed — added zero tap count guard. Returns error when transition_width is too large relative to sample_rate.

File: crates/sdr-dsp/src/taps.rs [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+
+    #[test]
+    fn test_estimate_tap_count_invalid() {
+        assert!(estimate_tap_count(0.0, TEST_SAMPLE_RATE).is_err());
+        assert!(estimate_tap_count(-1.0, TEST_SAMPLE_RATE).is_err());
+        assert!(estimate_tap_count(1_000.0, 0.0).is_err());
+        assert!(estimate_tap_count(f64::NAN, TEST_SAMPLE_RATE).is_err());
+        assert!(estimate_tap_count(f64::INFINITY, TEST_SAMPLE_RATE).is_err());
+        assert!(estimate_tap_count(f64::NEG_INFINITY, TEST_SAMPLE_RATE).is_err());
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/taps.rs:288]: Fixed — added test case: estimate_tap_count(1_000_000.0, 1_000.0) correctly returns error.
